### PR TITLE
feat(cache): implementation for cache mixin

### PR DESCRIPTION
**Pull request by barleendhaliwal**
## Description

- This package contains a mixin which can be used to cache get responses using redis.
- Added Readme file on how to use it and other configuration required

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [x] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/cache/src/mixins/cache.mixin.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+          }')
+          for _,key in ipairs(result[2]) do
+              redis.call('DEL', key)
+              dels = dels + 1
+          end
+          cursor = tonumber(result[1])
+      until cursor == 0
+      return dels`;
+      try {
+        return await this.executeRedisCommand(`EVAL`, [
```
Comment (samarpan-b) [inline packages/cache/src/mixins/cache.mixin.ts:127]: make sure this script is safe.

File: packages/cache/src/mixins/cache.mixin.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    checkDataSource() {
+      if (!this.cacheDataSource) {
+        throw new Error(`Please provide value for cacheDataSource`);
+      }
+    }
+
+    // returns promisified execute function
+    executeRedisCommand(command: string, args: (string | number)[]) {
+      return new Promise((resolve, reject) => {
```
Comment (samarpan-b) [inline packages/cache/src/mixins/cache.mixin.ts:147]: why wrapped in promise ?

File: packages/cache/src/mixins/cache.mixin.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        if (this.cacheDataSource.connector?.execute) {
+          // eslint-disable-next-line @typescript-eslint/no-floating-promises
+          this.cacheDataSource.connector.execute(
```
Comment (samarpan-b) [inline packages/cache/src/mixins/cache.mixin.ts:150]: this is already a promise. No need to wrap

File: packages/cache/src/mixins/cache.mixin.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    checkDataSource() {
+      if (!this.cacheDataSource) {
+        throw new Error(`Please provide value for cacheDataSource`);
+      }
+    }
+
+    // returns promisified execute function
+    executeRedisCommand(command: string, args: (string | number)[]) {
```
Comment (samarpan-b) [inline packages/cache/src/mixins/cache.mixin.ts:146]: this must be async function

File: packages/cache/src/mixins/cache.mixin.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      if (filter) {
+        key += `_${JSON.stringify(filter)}`;
+      }
+      if (options) {
+        key += `_${JSON.stringify(options)}`;
+      }
+      return key;
+    }
+
+    async clearCache() {
```
Comment (samarpan-b) [inline packages/cache/src/mixins/cache.mixin.ts:110]: mention return types please

File: packages/cache/src/mixins/cache.mixin.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+          cacheOptions.ttl ?? 60000,
+        ]);
+      } catch (err) {
+        throw new Error(
+          `Unexpected error occured while saving in cache : ${err}`,
+        );
+      }
+    }
+
+    getKey(id?: ID, filter?: FilterExcludingWhere<M>, options?: Options) {
```
Comment (samarpan-b) [inline packages/cache/src/mixins/cache.mixin.ts:96]: why this function ?

**Review by samarpan-b — CHANGES_REQUESTED** [review]
add test cases

File: packages/cache/src/mixins/cache.mixin.ts (hunk 2 above)
Comment (barleendhaliwal) [reply packages/cache/src/mixins/cache.mixin.ts:147]: If no callback function is given to execute function then it gives error : AssertionError [ERR_ASSERTION]: callback must be a function. Even if I give empty callback function and await the result will always be undefined.

File: packages/cache/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Check out https://hub.docker.com/_/node to select a new base image
```
Comment (akshatdubeysf) [inline packages/cache/Dockerfile:1]: purpose of docker file in a mixin repo?

**Review by akshatdubeysf — APPROVED** [review]

File: packages/cache/Dockerfile (hunk 7 above)
Comment (barleendhaliwal) [reply packages/cache/Dockerfile]: Deleted docker file

**Comment by barleendhaliwal** [comment]
Added Unit test cases 
Screenshot from 2022-04-29 14-49-16

**Pull request merged**